br/pkg/restore/import.go

```diff
@@ -866,6 +850,284 @@ func (importer *FileImporter) downloadRawKVSST(
+					if err != nil {
+						return err
+					}
+					return nil
+				}, utils.NewDownloadSSTBackoffer())
+				if err != nil {
+					return errors.Trace(err)
+				}
+				if resp.GetError() != nil {
+					return errors.Annotate(berrors.ErrKVDownloadFailed, resp.GetError().GetMessage())
+				}
+				if resp.GetIsEmpty() {
+					return errors.Trace(berrors.ErrKVRangeIsEmpty)
+				}
+
+				mu.Lock()
+				sstMeta, ok := downloadMetasMap[file.Name]
+				if !ok {
+					mu.Unlock()
+					return errors.New("not found file key for download sstMeta")
+				}
+				sstMeta.Range = &import_sstpb.Range{
+					Start: TruncateTS(resp.Range.GetStart()),
+					End:   TruncateTS(resp.Range.GetEnd()),
+				}
+				resultMetasMap[file.Name] = &sstMeta
+
+				log.Debug("download from peer",
+					logutil.Region(regionInfo.Region),
+					logutil.File(file),
+					logutil.Peer(peer),
+					logutil.Key("resp-range-start", resp.Range.Start),
+					logutil.Key("resp-range-end", resp.Range.End),
+					zap.Bool("resp-isempty", resp.IsEmpty),
+					zap.Uint32("resp-crc32", resp.Crc32),
+					zap.Int("len files", len(files)),
+				)
+				mu.Unlock()
+			}
+			return nil
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+	for _, sstMeta := range resultMetasMap {
+		downloadMetas = append(downloadMetas, sstMeta)
+	}
+	return downloadMetas, nil
+}
+
+func (importer *FileImporter) downloadRawKVSSTV2(
+	ctx context.Context,
+	regionInfo *split.RegionInfo,
+	files []*backuppb.File,
+	cipher *backuppb.CipherInfo,
+	apiVersion kvrpcpb.APIVersion,
+) ([]*import_sstpb.SSTMeta, error) {
+	downloadMetas := make([]*import_sstpb.SSTMeta, 0, len(files))
+	for _, file := range files {
+		uid := uuid.New()
+		id := uid[:]
+		// Empty rule
+		var rule import_sstpb.RewriteRule
+		sstMeta, err := GetSSTMetaFromFile(id, file, regionInfo.Region, &rule, RewriteModeLegacy)
+		if err != nil {
+			return nil, err
+		}
+
+		// Cut the SST file's range to fit in the restoring range.
+		if bytes.Compare(importer.rawStartKey, sstMeta.Range.GetStart()) > 0 {
+			sstMeta.Range.Start = importer.rawStartKey
+		}
+		if len(importer.rawEndKey) > 0 &&
+			(len(sstMeta.Range.GetEnd()) == 0 || bytes.Compare(importer.rawEndKey, sstMeta.Range.GetEnd()) <= 0) {
+			sstMeta.Range.End = importer.rawEndKey
+			sstMeta.EndKeyExclusive = true
+		}
+		if bytes.Compare(sstMeta.Range.GetStart(), sstMeta.Range.GetEnd()) > 0 {
+			return nil, errors.Trace(berrors.ErrKVRangeIsEmpty)
+		}
+
+		req := &import_sstpb.DownloadRequest{
+			Sst:            *sstMeta,
+			StorageBackend: importer.backend,
+			Name:           file.GetName(),
+			RewriteRule:    rule,
+			IsRawKv:        true,
+			CipherInfo:     cipher,
+			StorageCacheId: importer.cacheKey,
+		}
+		log.Debug("download SST", logutil.SSTMeta(sstMeta), logutil.Region(regionInfo.Region))
+
+		var atomicResp atomic.Value
+		eg, ectx := errgroup.WithContext(ctx)
+		for _, p := range regionInfo.Region.GetPeers() {
+			peer := p
+			eg.Go(func() error {
```

Comment:
It seems there isn't a concurrency limit. Is it possible to extract more common logic (like handling the response and resource control, etc..) to common functions?

Reply:
no limit for raw kv sst here is expected.  we could do it later